### ISSUE TITLE
ccache randomly fails; also please add cmake option disabling ccache

### ISSUE THREAD

**Issue by yurivict** [qTox/qTox#4753]
ccache randomly fails on the FreeBSD port, version 1.12.1:
```
ccache: error: /usr/ports/net-im/qTox/work/.ccache/ccache.conf: No such file or directory
```
It doesn't fail every time, only sometimes.

Additionally, please add the the option that can disable it, ex.:
```
option(USE_CCACHE "Use ccache to speed up compile" ON)
```

**Comment by anthonybilinski** [comment]
ccache has a built in option to disable it:

> disable (CCACHE_DISABLE) [boolean]
>           When true, ccache will just call the real compiler, bypassing the cache completely. The default is false.

I think this means that adding an option to disable it from qTox's build system isn't really necessary.

As for the ccache error, it seems like a generic ccache problem. Do you think it's caused by qTox, or is it just the reason for wanting to disable it?

**Comment by yurivict** [comment]
> As for the ccache error, it seems like a generic ccache problem.

I think so.

> ccache has a built in option to disable it:
> I think this means that adding an option to disable it from qTox's build system isn't really necessary.

CCACHE_DISABLE is an environment variable. Setting it doesn't disable ccache for me for some reason.

These lines
```
if(CCACHE_FOUND)
  set_property(GLOBAL PROPERTY RULE_LAUNCH_COMPILE ccache)
  message(STATUS "using ccache")
endif()
```
ccache runs regardless of ```CCACHE_DISABLE```. This is why ```USE_CCACHE``` IMO is needed.

**Comment by anthonybilinski** [comment]
Setting it at cmake time wouldn't help since qTox would still find ccache and call it on every build, but setting CCACHE_DISABLE environment variable at build time will cause ccache would pass through directly, so wouldn't actually use itself, would just call gcc.

i.e.
```
cmake .
CCACHE_DISABLE=1 make 
```

If ccache is ignoring CCACHE_DISABLE even at build time, or if it's failing before it checks the environment variable (possible if it's trying to read settings out of a file first) then I guess it can only be disabled from qTox. Adding the option to qTox's build system has the benefit of not having to CCACHE_DISABLE=1 on every build, if for some reason you want ccache with other projects but not with qTox.

**Comment by yurivict** [comment]
> Setting it at cmake time wouldn't help since qTox would still find ccache

No, a cmake variable will prevent it altogether.

> CCACHE_DISABLE=1 make

make is never called in the FreeBSD port case. FreeBSD uses ninja instead of make. Not sure how does ```CCACHE_DISABLE``` work in ninja.

**Comment by anthonybilinski** [comment]
I agree that a qTox cmake variable could prevent it altogether, I meant that setting CCACHE_DISABLE at cmake time wouldn't do anything, since we don't look for it.

No matter how ccache is called, it disables itself with the CCACHE_DISABLE environment variable. It's not up to the caller. `make` doesn't handle the variable at all, it just calls ccache as normal, which immediately passes through once it checks the variable. So even with ninja, CCACHE_DISABLE should still work correctly.

**Comment by yurivict** [comment]
> No matter how ccache is called, it disables itself with the CCACHE_DISABLE environment variable.

This clearly doesn't work this way. Build still fails when ```CCACHE_DISABLE=1```.

**Comment by anthonybilinski** [comment]
That's at least the behaviour ccache says it has :S https://ccache.samba.org/manual.html#_boolean_values

Maybe it's hitting the error of not finding the file before it checks the disable variable. I guess it doesn't hurt to add a cmake option to disable it. I just can't think of any reason it should be needed except for when the system ccache isn't working, and then why have ccache installed at all.

**Comment by yurivict** [comment]
> Maybe it's hitting the error of not finding the file

Exactly. ccache is obviously buggy. Instead of it helping us, we have to be working on their bugs.
I only need to have a stable qTox port. I don't particularly care for ccache being used, because qTox builds fine without it. Bugs in ccache negate all its benefit as it is right now.

This is why there should be a way to reliably disable ccache until it is stable, like I suggested previously.

**Comment by sudden6** [comment]
I think an option to disable ccache wouldn't be to bad, since it's an optional dependency and we have switches for every optional dependency.

@yurivict why not uninstall ccache in the mean time if it's buggy? then cmake won't detect it and won't use it.

**Comment by yurivict** [comment]
> why not uninstall ccache in the mean time if it's buggy? then cmake won't detect it and won't use it.

Because I am not talking about myself, but about the port. Port can be built on any computer, some will have ccache preinstalled. This means that qTox conflicts with ccache. I will need to patch CMakeFile.txt to delete ccache to prevent this. So I am asking to have such ability upstream, in the qTox sources.

**Comment by yurivict** [comment]
https://github.com/qTox/qTox/pull/4982

**Comment by sudden6** [comment]
Fixed by #4982
